### ISSUE TITLE
[Feature] Log Browsing History over multiple sessions.

### ISSUE THREAD

**Issue by Brobin** [michael-lazar/rtv#81]
I love the feature that changes the color of visited links and posts in one session. It would be nice to have this work across sessions. Perhaps a history log file that is loaded on program start into the `opened_links` variable.

```
[rtv]
history=True
```

Then save in in a file `rtv_history.log` etc.

**Comment by Brobin** [comment]
One issue I think I will run into is having a very large file. Perhaps links older than 1 month are not saved

**Comment by Brobin** [comment]
Also, I'm not too sure about log file standards. Would it be okay to put it in `.config/rtv/history.log`?

**Comment by michael-lazar** [comment]
From what I can find it either needs to go in ``XDG_CACHE_HOME`` or ``XDG_DATA_HOME``. It's unclear to me which one is correct, but I would lean towards CACHE because it's non-essential data. I would also name the file ``history.txt`` because it's not a standard log in the sense that it will be consumed and overwritten by the program.

Using timestamps for each link sounds like a lot of overhead, it might be better to just save the most recent N links.

**Comment by Brobin** [comment]
To save the most recent N links, we would have to use some sort of ordered list. A set is unordered so restricting the size of it when writing would not guarantee that you would get the last N links, just N random links.

Would you be opposed to refactoring to use a list instead? That way we could just slice the end of the list for the output file.

Also, is 1000 a reasonable number of links to save?

**Comment by michael-lazar** [comment]
I used a set because testing for membership is O(1) as opposed to lists which are O(N). I don't think going back to a list is a good idea. Also think about how you're going to write to this file. If you only write when the program is quit via ``q``, you will be losing history on ``ctrl-c``.

Here's one approach that uses file appends to avoid having to overwrite the whole file after each time. There may be other approaches as well.

1. Read the file in as a list.
2. If the list length is over N, overwrite the file with only the most recent N links.
3. Convert the list to a set
4. When a new link is clicked, add it to the set AND append the link directly to the end of the file.

As for the number of links, I think 100-200 would probably be enough.

**Comment by Brobin** [comment]
Good point about the complexity. I like this approach. It has lot of extra file operations, but I guess that's necessary to not lose history if you quit with `ctrl+c`.

**Comment by noahmorrison** [comment]
As for the ```ctrl+c``` problem, why not use ```atexit```?

```python
import atexit
import time

@atexit.register
def save():
    print('saving')

if __name__ == '__main__':
    time.sleep(2)
    print('done')
```

run this, no matter what (ctrl+c, or normal exit), "saving" will always be printed.

**Comment by Brobin** [comment]
Didn't know that module existed! Sounds good to me. If we use that, I think it would be best to do the file I/O at exit instead of when you visit each link. It would keep the code cleaner.

**Comment by Brobin** [comment]
I've been thinking about this and the implementation that makes the most sense to me is a stack. Here's my proposal:

1. Load the file into a list (stack)
2. Add every link to that list during the session (don't even check for membership, if you visit an element again, it goes back on top of the stack)
3. On exit, pop the last N elements off of the list (using `stack.pop()`)
4. Write those links to the file.

Pros:
 - Order is maintained
 - Fewer file operations. One on launch and one on exit

Cons:
 - May be duplicate elements
 - Will not save history if they close the terminal

**Comment by Brobin** [comment]
I've almost got a full implementation working. @michael-lazar, it is very similar to how you described it, except it uses the a`@atexit` decorator to do one file operation at exit instead of loading and writing to the file all of the time.

**Comment by michael-lazar** [comment]
Merged in #95
